src/main.rs

```diff
@@ -495,3 +512,38 @@ fn main_recv(addr: &str, n_conn: &str) -> Result<()> {
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{thread, time::Duration};
+
+    #[test]
+    fn test_accepts_valid_protocol() {
+        thread::spawn(|| {
+            std::fs::File::create("a-file").unwrap();
+            main_send("127.0.0.1:1234", &["a-file".into()], 1).unwrap();
+        });
+        thread::sleep(Duration::from_millis(100));
```

Comment:
Similar to the previous comment, not great, but gives the TcpListener time to bind to the provided address.

Reply:
Yeah I always try very hard to avoid sleeps in tests, it’s only a proxy for the true readiness so when the thing is ready early it makes the tests unnecessarily slow, and when the thing takes longer to be ready then the test fails. (Which locally never happens, but two years later we have many more tests and Cargo runs them in parallel, and then it runs on an underpowered machine on CI and 1 in 20 times it’s unlucky and it times out, and now we have a flaky test …)

If we want to test on this level at all, then I think the proper solution here is to create a [`Condvar`](https://doc.rust-lang.org/std/sync/struct.Condvar.html), let the sender signal it after the `bind` but before the `accept`, and then here instead of sleeping we wait for the cond var. It’s not much worse than the sleep, just one more argument to the main function.

(Actually now that I’m looking at it, `CondVar` has this business with the mutex that we don’t need, we could also use a [`Barrier`](https://doc.rust-lang.org/std/sync/struct.Barrier.html) or [`channel<()>`](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html).)